std/typecons.d

```diff
@@ -2188,6 +2199,320 @@ Rebindable!T rebindable(T)(Rebindable!T obj)
     assert(rebindable(pr3341_aa)[321] == 543);
 }
 
+/** Models safe reassignment of otherwise constant struct instances.
+ *
+ * A struct with a field of reference type cannot be assigned to a constant
+ * struct of the same type. `Rebindable!(const S)` allows assignment to
+ * a `const S` while enforcing only constant access to its fields.
+ *
+ * `Rebindable!(immutable S)` does the same but field access may create a
+ * temporary copy of `S` in order to enforce _true immutability.
+ */
+template Rebindable(S)
+if (is(S == struct))
+{
+    static if (isMutable!S)
+        alias Rebindable = S;
+    else
+    struct Rebindable
+    {
+    private:
+        // mutPayload's pointers must be treated as tail const
+        void[S.sizeof] mutPayload;
+
+        void emplace(ref S s)
+        {
+            import std.conv : emplace;
+            static if (__traits(compiles, () @safe {S tmp = s;}))
+                () @trusted {emplace!S(mutPayload, s);}();
+            else
+                emplace!S(mutPayload, s);
+        }
+
+    public:
+        this()(auto ref S s)
+        {
+            emplace(s);
+        }
+
+        // immutable S cannot be passed to auto ref S above
+        static if (!is(S == immutable))
+        this()(immutable S s)
+        {
+            emplace(s);
+        }
+
+        void opAssign()(auto ref S s)
+        {
+            movePayload;
+            emplace(s);
+        }
+
+        static if (!is(S == immutable))
+        void opAssign()(immutable S s)
+        {
+            movePayload;
+            emplace(s);
+        }
+
+        void opAssign(Rebindable other)
+        {
+            this = other.trustedPayload;
+        }
+
+        // must not escape when S is immutable
+        private ref trustedPayload() @trusted
+        {
+            return *cast(S*) mutPayload.ptr;
+        }
+
+        static if (!is(S == immutable))
+        private ref S Rebindable_getRef() @property
+        {
+            // payload exposed as const ref when S is const
+            return trustedPayload;
+        }
+
+        static if (is(S == immutable))
+        private S Rebindable_get() @property
+        {
+            // we return a copy for immutable S
```

Comment:
@wilzbach re: https://github.com/dlang/phobos/pull/6136#issuecomment-364078652
>However, if the getRef code is used for immutable, no temporary variable would be created and thus resulting in i.e. the postblit not being called

I phrased my question unclearly. Why is this necessary when `S == immutable` if it's not necessary when `S == const`?